### PR TITLE
[SPARK-40834][SQL][FOLLOWUP] Take care of legacy query end events

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/38302 . For events generated by old versions of Spark, which do not have the new `errorMessage` field, we should use the old way to detect query execution status (failed or not).

This PR also adds a UI test for the expected behavior.

### Why are the changes needed?

backward compatibility

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala [hunk 1]
```diff
@@ -489,13 +489,13 @@ private class LiveExecutionData(val executionId: Long) extends LiveEntity {
...
   var submissionTime = -1L
   var completionTime: Option[Date] = None
+  var errorMessage: Option[String] = None
 
   var jobs = Map[Int, JobExecutionStatus]()
   var stages = Set[Int]()
   var driverAccumUpdates = Seq[(Long, Long)]()
 
   @volatile var metricsValues: Map[Long, String] = null
-  var errorMessage: Option[String] = None
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala:498]: nit: move `errorMessage` closer to `completionTime`, as they both indicate the end of the execution.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusStore.scala [hunk 2]
```diff
@@ -90,6 +90,7 @@ class SQLExecutionUIData(
     val metrics: Seq[SQLPlanMetric],
     val submissionTime: Long,
     val completionTime: Option[Date],
+    val errorMessage: Option[String],
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusStore.scala:93]: ditto

**Review by ulysses-you — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 3]
```diff
@@ -121,7 +121,11 @@ object SQLExecution {
               SparkThrowableHelper.getMessage(e)
           }
           val event = SparkListenerSQLExecutionEnd(
-            executionId, System.currentTimeMillis(), errorMessage)
+            executionId,
+            System.currentTimeMillis(),
+            // Use empty string to indicate no error, as None may mean events generated by old
+            // versions of Spark.
+            errorMessage.orElse(Some("")))
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:128]: good point

**Comment by cloud-fan** [comment]
thanks for review, merging to master!

**Pull request closed without merging**